### ISSUE TITLE
Please add install and basic usage instructions in README

### ISSUE THREAD

**Issue by rgommers** [FFY00/mesonpy#20]
`pip install mesonpy` gives:
```
$ pip install mesonpy 
Collecting mesonpy
  Using cached MesonPy-0.0.1.tar.gz (14 kB)
  Preparing metadata (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /home/rgommers/anaconda3/envs/scipy-dev/bin/python3.9 -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-m_0m1w7a/mesonpy_f4910b5198f8497aa8f6fa1184381e91/setup.py'"'"'; __file__='"'"'/tmp/pip-install-m_0m1w7a/mesonpy_f4910b5198f8497aa8f6fa1184381e91/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-4xq7gbqy
       cwd: /tmp/pip-install-m_0m1w7a/mesonpy_f4910b5198f8497aa8f6fa1184381e91/
  Complete output (5 lines):
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "/tmp/pip-install-m_0m1w7a/mesonpy_f4910b5198f8497aa8f6fa1184381e91/setup.py", line 4, in <module>
      with open('requirements.txt') as f:
  FileNotFoundError: [Errno 2] No such file or directory: 'requirements.txt'
  ----------------------------------------
WARNING: Discarding https://files.pythonhosted.org/packages/e5/b6/ff8146b0332f23fc2c242a9e20204ef217249764c06eebac301609b21426/MesonPy-0.0.1.tar.gz#sha256=27a81c9af4f5bdea934bca0ac698a85f44ca6b26a9eaa096dc0f1fbf22108739 (from https://pypi.org/simple/mesonpy/). Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
ERROR: Could not find a version that satisfies the requirement mesonpy (from versions: 0.0.1)
ERROR: No matching distribution found for mesonpy
```

**Comment by rgommers** [comment]
Well duh, because the PyPI name doesn't match the package name: https://pypi.org/project/meson-python/

Guess the request is for some docs (in the README for now) - how to install, and how to enable it (easy, but now has to be derived from test cases in this repo).

**Comment by rgommers** [comment]
In test cases in this repo, it uses `requires = 'mesonpy'` - this is a problem, if the package is named differently on PyPI.

**Comment by rgommers** [comment]
Related: how does one build a very basic wheel in a local dev env. The invocation I want, I think, is:
```
$ python -m build --wheel --no-isolation --skip-dependency-check
```
I'm not sure what the `pip` equivalent is for `--skip-dependency-check`.

**Comment by rgommers** [comment]
Then next one I run into when building a SciPy wheel:
```
mesonpy.MesonBuilderError: Unsupported dynamic fields: description
```
Which is fine, can get rid of that. Doing that gives:
```
pep621.ConfigurationError: Missing version field
```
The version is dynamic too, since we encode the git commit hash in it for dev builds. But maybe wheel rename can/should be done separately after building it, because Meson cannot have the `project-version` field be dynamic. We obtain it post build from `scipy.__version__`, which comes from an autogenerated file `scipy/version.py`.

**Comment by rgommers** [comment]
Next issue:
```
FileNotFoundError: [Errno 2] No such file or directory: 'auditwheel'
```
Where is the flag to skip `auditwheel` completely? I do not want, need or like `auditwheel` for a local install.

**Comment by rgommers** [comment]
Installed `auditwheel`, then ran into gh-12:
```
auditwheel: error: cannot repair "/tmp/mesonpy-ymwx6fyb/build/scipy-1.9.0.dev0-py3-none-linux_x86_64.whl" to "manylinux_2_5_x86_64" ABI because of the presence of too-recent versioned symbols. You'll need to compile the wheel on an older toolchain
```
That seems to be the end of the line here for now. There's no good way to get past this as far as I can tell. 

We need to think about UX a bit as well:
- `python -m build` by default builds a wheel which is suitable for redistribution on PyPI in an isolated env. 
- We need a simple way to say _"build me a wheel matching my local env to install it"_. Right now it's something like `python -m build --wheel --no-isolation --skip-dependency-check --new-arg-to-skip-audithweel`. Which is unusable.

**Comment by rgommers** [comment]
Cc @matthew-brett for visibility.

**Comment by rgommers** [comment]
For completeness, the whole diff for current SciPy master is:
```diff
diff --git a/pyproject.toml b/pyproject.toml
index d3a01ecfe..72c89f613 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -8,8 +8,10 @@
 #     "pybind11>=2.4.3,<2.5.0",
 
 [build-system]
+build-backend = 'mesonpy'
 requires = [
     "wheel",
+    "meson-python",
     "setuptools<60.0",  # do not increase, 60.0 enables vendored distutils
     "Cython>=0.29.18",
     "pybind11>=2.4.3",
@@ -45,7 +47,7 @@ requires = [
 [project]
 name = "SciPy"
 license = {file = "LICENSE.txt"}
-
+version = "1.9.0.dev0"
 maintainers = [
     {name = "SciPy Developers", email = "scipy-dev@python.org"},
 ]
@@ -76,7 +78,6 @@ classifiers = [
     "Operating System :: Unix",
     "Operating System :: MacOS",
 ]
-dynamic = ['version', 'description']
 ```

**Comment by FFY00** [comment]
Hi, thank you for filling this, and sorry for the lack of documentation. I will add the missing `skip-external-bundling` option (#6), which is not yet exposed because we rely on auditwheel for the bundling of internal libraries, and document the config options soon.

**Comment by rgommers** [comment]
> because we rely on auditwheel for the bundling of internal libraries

I'm not sure I understand this. You only need this for building a wheel for uploading for PyPI, and not for local use, redistribution on a private index where you control the system (e.g. a single Docker image), for Linux distro packaging, or anything else. Therefore I believe `auditwheel` should be disabled by default, and is not necessary for initial testing.

**Comment by FFY00** [comment]
In your Meson project, you might build libraries and then link them against the native Python modules you build, we still need to vendor those libraries. Currently, we are relying on auditwheel for that, but I have been working on moving away from it and doing it ourselves.

**Comment by rgommers** [comment]
Ah, interesting. That's a pretty unusual pattern though. Not used in SciPy. We typically do one of these:

- simplest: build a Python extension module (a `.so`)
- more complex: build a static library, then link that into another Python extension module

Your case should be automatically handled as well I'd expect. The wheel should always contain everything that is installed by `meson install`. So if your standalone library isn't installed, then a direct install via `meson install` is broken. And hence it's a user bug, you should not need to take the extra action of inspecting Python extension modules to see what they use and then vendor that.

**Comment by rgommers** [comment]
I looked at this use case some more, since it still wasn't completely clear what the idea was. I re-read the wheel spec to make sure that I was correct in assuming that a wheel cannot contain anything that does not go into `site-packages` (that is indeed correct). So what's left is some non-Python standalone library being installed as a random static or dynamic library into `site-packages` somewhere. Which is covered, I think, by a test case like https://github.com/FFY00/mesonpy/tree/main/tests/packages/library-pep621.

It's unclear to me that this is a real-world use case. It certainly seems incomplete, because how is another package supposed to use such a library? There is no good way to discover it (neither a Python API to query, nor `pkg-config` support, not on a default library search path), so all one can do is just hardcoding the path. One library I could think of that does such a thing is MKL. It unpacks its files to `mkl-2022.0.1.data` (checked by downloading and unpacking wheel for latest version):
```
ls mkl-2022.0.1.data/data 
lib
(base) % ls mkl-2022.0.1.data/data/lib/
libmkl_avx.so.2			libmkl_gnu_thread.so.2		libmkl_p4m3.so.2		libmkl_vml_avx2.so.2		libmkl_vml_p4m.so.2
libmkl_avx2.so.2		libmkl_intel.so.2		libmkl_rt.so.2			libmkl_vml_avx512.so.2		libmkl_vml_p4m2.so.2
libmkl_avx512.so.2		libmkl_intel_thread.so.2	libmkl_sequential.so.2		libmkl_vml_cmpt.so.2		libmkl_vml_p4m3.so.2
libmkl_core.so.2		libmkl_p4.so.2			libmkl_tbb_thread.so.2		libmkl_vml_ia.so.2		mkl_msg.cat
libmkl_gf.so.2			libmkl_p4m.so.2			libmkl_vml_avx.so.2		libmkl_vml_p4.so.2
```
So you cannot really special `requires = 'mkl'` and do anything with the result of that, the path is a mess. If it had gone into `mkl/` then at least you could hardcode the path from your own package as `../mkl/data/` at runtime. Which will require some platform-dependent library search path manipulation too (e.g., https://github.com/MacPython/scipy-wheels/blob/master/_distributor_init.py#L25).

Another thought re `auditwheel`: it's not cross-platform, so not all that useful for this kind of vendoring. For internal libraries, vendoring needs to work the same way across OSes; the only thing that is a little platform-specific is library search path changes (needs special-casing for Windows at least).

**Comment by FFY00** [comment]
> I looked at this use case some more, since it still wasn't completely clear what the idea was. I re-read the wheel spec to make sure that I was correct in assuming that a wheel cannot contain anything that does not go into `site-packages` (that is indeed correct).

It can also include headers, script, and general data, in the [`.data` directory](https://www.python.org/dev/peps/pep-0427/#the-data-directory).

> So what's left is some non-Python standalone library being installed as a random static or dynamic library into `site-packages` somewhere. Which is covered, I think, by a test case like [`main`/tests/packages/library-pep621](https://github.com/FFY00/mesonpy/tree/main/tests/packages/library-pep621?rgh-link-date=2022-01-04T05%3A33%3A43Z).

Yes, exactly. It is covered by that test case but I should have a separate case that builds a library like that and uses links it in a native Python module.

> It's unclear to me that this is a real-world use case. It certainly seems incomplete, because how is another package supposed to use such a library? There is no good way to discover it (neither a Python API to query, nor `pkg-config` support, not on a default library search path), so all one can do is just hardcoding the path. One library I could think of that does such a thing is MKL. It unpacks its files to `mkl-2022.0.1.data` (checked by downloading and unpacking wheel for latest version):

I think a good real world use-case example would be something like https://www.kicad.org/ (it uses CMake but let's pretend for a bit it used Meson), which is an EDA application (used to design circuits, PCBs, etc.) that allows Python scripting. The main target of the project is not a Python package, but it does expose a Python module to enable scripting. So, the build system will be written with normal shared libraries.
With this support for scripting, you start seeing projects like https://github.com/yaqwsx/PcbDraw and https://github.com/yaqwsx/KiKit to pop up, which you will want to run on CI or readthedocs for eg. Being able to build a wheel for kicad would give users a much easier way to get that kicad Python module into the target system -- currently, I think that would be the only way to get it to work on readthedocs.

SciPy is specifically designed as a Python package, so it is understandable that it will use static libraries, but I think any project that isn't designed primarily as a Python package probably won't do that.

Anyway, we can ignore that for a bit right now as we are working on this for SciPy primarily.

> So you cannot really special `requires = 'mkl'` and do anything with the result of that, the path is a mess. If it had gone into `mkl/` then at least you could hardcode the path from your own package as `../mkl/data/` at runtime. Which will require some platform-dependent library search path manipulation too (e.g., [MacPython/scipy-wheels@`master`/_distributor_init.py#L25](https://github.com/MacPython/scipy-wheels/blob/master/_distributor_init.py?rgh-link-date=2022-01-04T05%3A33%3A43Z#L25)).

That is true. Something like that could be done with custom support in the build system, but it requires packages to play into a new specification, essentially.

**Comment by rgommers** [comment]
> Being able to build a wheel for kicad would give users a much easier way to get that kicad Python module into the target system

There's multiple issues with this though, both technical and social. The PyPI model is author-led and very ad-hoc; if the KiCad authors themselves have a good reason to package their own project on PyPI then they should of course do that. But I suspect they won't want that or care - and another project doing that should be strongly discouraged, and certainly for making life on RTD easier is not a good enough reason to do this. What you're then going to have is one random person/project taking the `kicad` name on PyPI (hypothetical, it's already being name-squatted: https://pypi.org/project/kicad/), packaging it just for the platform and Python version they need (Ubuntu I guess for RTD), and leaving it at that.

RTD is also special, in that it's so restricted that you can't do anything but upload a requirements file. In pretty much any other circumstance, `curl` would probably be a better fit than `pip` here if you just want to download and unpack some random binary and put it on the PATH.

I am contemplating doing exactly this "package a non-Python package" for OpenBLAS, but am really hesitant because it's an unhealthy idea (unfortunately, we may be forced to at some point).

tl;dr yes there's a possible use case, but we shouldn't encourage it.

**Comment by FFY00** [comment]
> There's multiple issues with this though, both technical and social. The PyPI model is author-led and very ad-hoc; if the KiCad authors themselves have a good reason to package their own project on PyPI then they should of course do that.

I agree. At the end of the day, we still require a `pyproject.toml` configured with the backend, so it is very much something that should be done by the original developers.

Let me give you another, perhaps better, example: libsolv, which is the dependency resolution backend we are integrating into Conda. The upstream provides Python bindings, but does not release them on PyPI, probably because the build system just does not support that -- they can't easily build a wheel to distribute. This is not a big deal in our case in Conda because Conda is able to manage this native dependency, but let's say we are on pip and wanted to do the same thing we are doing in Conda, we would either need the libsolv Python module to be released on PyPI, or we would have to ask users to install libsolv with the Python bindings enabled on their system, which it's not a good UX.
If they moved to Meson and used this backend, they could have the native build system build the Python module just like they are right now, but also be able to build a wheel for the Python bindings for distribution. Note that this wheel would only make available the Python module, not any libraries built by the project.

> I am contemplating doing exactly this "package a non-Python package" for OpenBLAS, but am really hesitant because it's an unhealthy idea (unfortunately, we may be forced to at some point).

Our goal here would not be to package libraries themselves, but Python modules that used those libraries. Internally, we would build the actual library and embedded it in the wheel for the Python module to use, but that would be a private detail of the wheel.
I think that packaging libraries themselves could indeed be unhealthy, but not what I described, which is already happening.

Did you get my point now :sweat_smile:? Hopefully, I was able to clarify the intent, and how this would work, here.

**Comment by rgommers** [comment]
This makes sense, and `libsolv` is a nice example, thanks.

**Comment by LukeLIN-web** [comment]
I also meet this problem,  
```
pip3 install mesonpy
Collecting mesonpy
  Using cached MesonPy-0.0.1.tar.gz (14 kB)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [7 lines of output]
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/tmp/pip-install-0o3x3b9d/mesonpy_3fd0dc16801a4805af66479d061250f9/setup.py", line 4, in <module>
          with open('requirements.txt') as f:
               ^^^^^^^^^^^^^^^^^^^^^^^^
      FileNotFoundError: [Errno 2] No such file or directory: 'requirements.txt'
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```
